components/PinchZoom/index.ts

```diff
@@ -0,0 +1,274 @@
+import './styles.css';
+import { PointerTracker, Pointer } from '../../utils/trackDragging';
+
+interface Point {
+  clientX: number;
+  clientY: number;
+}
+
+function getDistance (a: Point, b?: Point): number {
+  if (!b) return 0;
+  return Math.sqrt((b.clientX - a.clientX) ** 2 + (b.clientY - a.clientY) ** 2);
+}
+
+function getMidpoint (a: Point, b?: Point): Point {
+  if (!b) return a;
+
+  return {
+    clientX: (a.clientX + b.clientX) / 2,
+    clientY: (a.clientY + b.clientY) / 2
+  };
+}
+
+// I'd rather use DOMMatrix/DOMPoint here, but the browser support isn't good enough.
+// Given that, better to use something everything supports.
+let cachedSvg: SVGSVGElement;
+
+function getSVG (): SVGSVGElement {
+  return cachedSvg || (cachedSvg = document.createElementNS('http://www.w3.org/2000/svg', 'svg'));
+}
+
+function createMatrix (): SVGMatrix {
+  return getSVG().createSVGMatrix();
+}
+
+function createPoint (): SVGPoint {
+  return getSVG().createSVGPoint();
+}
+
+export default class PinchZoom extends HTMLElement {
+  // The element that we'll transform.
+  // Ideally this would be shadow DOM, but we don't have the browser
+  // support yet.
+  private _positioningEl?: HTMLElement;
+  // Current transform.
+  private _x = 0;
+  private _y = 0;
+  private _scale = 1;
+
+  constructor () {
+    super();
+
+    // Watch for children changes.
+    // Note this won't fire for initial contents,
+    // so _stageElChange is also called in connectedCallback.
+    new MutationObserver(() => this._stageElChange())
+      .observe(this, { childList: true });
+
+    // Watch for pointers
+    const pointerTracker: PointerTracker = new PointerTracker(this, {
+      start: (pointer, event) => {
+        // We only want to track 2 pointers at most
+        if (pointerTracker.currentPointers.length === 2) return false;
+        event.preventDefault();
+        return true;
+      },
+      move: previousPointers => {
+        this._onPointerMove(previousPointers, pointerTracker.currentPointers);
+      }
+    });
+
+    this.addEventListener('wheel', event => this._onWheel(event));
+  }
+
+  connectedCallback () {
+    this._stageElChange();
+  }
+
+  get x () {
+    return this._x;
+  }
+
+  get y () {
+    return this._y;
+  }
+
+  get scale () {
+    return this._scale;
+  }
+
+  /**
+   * Update the stage with a given scale/x/y.
+   */
+  setTransform (scale: number = this._scale, x: number = this._x, y: number = this._y, {
+    allowChangeEvent = false
```

Comment:
I thought the same, but doesn't `fireChangeEvent` make it sound like it *will* fire a change event, whereas in reality it might not, if the provided values match the current state, or are clamped to be equal to the current state?

Reply:
I think the word “change” covers that.

Reply:
@developit @kosamari what do you think?

```js
pinchZoom.setTransform(1, 100, 100, {
  allowChangeEvent: true
});
```

Or `fireChangeEvent`?

If you set this property to true, it may not fire a change event, eg if the supplied or computed values are the same as the current transform.